### PR TITLE
Implement SFTP Source

### LINKED ISSUE #67: Implement Sftp source
including multi-source, metadata-store, and different file access modes (list only, contents, line)

### PR DESCRIPTION
**Pull request by dturanski**
Resolves #67

### CODE REVIEW

File: functions/supplier/sftp-supplier/src/test/resources/logback.xml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright 2020-2020 the original author or authors.
```
Comment (artembilan) [inline functions/supplier/sftp-supplier/src/test/resources/logback.xml:3]: May we not have Copyright in the XML config?

File: functions/supplier/sftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/sftp/SftpSupplierConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+		chainFilter.addFilter(new SftpPersistentAcceptOnceFileListFilter(metadataStore, METADATA_STORE_PREFIX));
+		return chainFilter;
+	}
+
+	/*
+	 * Create a Flux from a MessageSource that will be used by the supplier.
+	 */
+	private Flux<Message<?>> sftpMessageFlux(MessageSource sftpMessageSource,
+			SftpSupplierProperties sftpSupplierProperties) {
+		return Mono
```
[the remark this answers is not in the document]
Comment (dturanski) [reply functions/supplier/sftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/sftp/SftpSupplierConfiguration.java:166]: I don't understand what the change is.  Can you give me the whole thing?

[the remark this answers is not in the document]
Comment (artembilan) [reply functions/supplier/sftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/sftp/SftpSupplierConfiguration.java:166]:
```
private Flux<Message<?>> sftpMessageFlux(MessageSource sftpMessageSource,
			SftpSupplierProperties sftpSupplierProperties) {

    return IntegrationReactiveUtils.messageSourceToFlux(messageSource)
	                .subscriberContext((ctx) ->
                 		ctx.put(IntegrationReactiveUtils.DELAY_WHEN_EMPTY_KEY, sftpSupplierProperties.getDelayWhenEmpty()))
}
```

File: functions/supplier/sftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/sftp/SftpHeaders.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.fn.supplier.sftp;
+
+/**
+ * @author David Turanski
```
[the remark this answers is not in the document]
Comment (dturanski) [reply functions/supplier/sftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/sftp/SftpHeaders.java:20]: I can do you one better. It no longer has a purpose. Good catch

File: functions/supplier/sftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/sftp/SftpSupplierConfiguration.java [hunk 4]
```diff
@@ -107,7 +107,7 @@
 	@Bean
 	@Primary
 	public MessageSource sftpMessageSource(
```
Comment (artembilan) [inline functions/supplier/sftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/sftp/SftpSupplierConfiguration.java:109]: You are still missing `<?>` over here

**Comment by artembilan** [comment]
Merged as https://github.com/spring-cloud/stream-applications/commit/8cc6f546b659a915dd41ac81aee6cbf05dcd6e78 after fixing more generics and making one of the test OS-independent.

**Pull request closed without merging**